Fix compiler warnings

Fireball-Teeworlds opened the pull request:
Not a single compiler warning remaining! (Until some future Rust release that is... :) ).

One other thing I did was remove 1.34.2 from github workflow matrix - it was preventing me from replacing `mem::uninitialized()` (deprecated in 1.39.0) with `mem::MaybeUninit` (added in 1.36.0).

heinrich5991 commented on .github/workflows/build.yaml:
```diff
@@ -12,7 +12,6 @@ jobs:
         rust:
           - nightly
           - stable
-          - 1.34.2
```
Rust 1.34.2 is the newest version supported in Debian stable: https://packages.debian.org/buster/rustc. I want the crate to be compilable there, hence I'd prefer not to drop support for it.

To me, mechanically replacing `mem::uninitialized()` with `mem::MaybeUninit::uninit().assume_init()` does not seem worth bumping the minimum supported Rust version. If you want to achieve a warning-free build, we can create a `common::mem::uninitialized` function that just calls `mem::uninitialized()` but has an `#[allow(deprecated)]` annotation to suppress the warning.

Fireball-Teeworlds replied:
Do we want to support rustc from Debian stable? I think given that rust
a) is being actively developed b) has it's own tool for installing the latest version - rustup c) lists this tool as the only installation method on their website - we could consider stepping away from supporting the packaged version.

But I'm also ok with keeping 1.34.2 since that was admittedly the only place where we've encountered a problem yet.

Btw, what do you think about removing those uninitialized buffers completely (making them initialized)? According to [rust doc](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html), we get undefined behaviour even with i32:

```
use std::mem::{self, MaybeUninit};

let x: i32 = unsafe { mem::uninitialized() }; // undefined behavior! ⚠️
// The equivalent code with `MaybeUninit<i32>`:
let x: i32 = unsafe { MaybeUninit::uninit().assume_init() }; // undefined behavior! ⚠️
```
I suspect that the overhead from initialising buffers in our case would be negligible.

def- replied:
For reference: We use Debian stable on all DDNet servers. Seems like a popular enough distribution to support.

Fireball-Teeworlds replied:
@def- we could use rustup to install rustc. It's listed as the official installation method so there shouldn't be many problems with it. I've just switched to it on my ubuntu desktop to install nightly toolchain for https://github.com/google/rerast. It's capable of keeping arbitrary number of toolchains installed for different purposes.

But again I'm also ok with keeping 1.34.2 supported for now since we only encountered one issue with it.

heinrich5991 replied:
> Do we want to support rustc from Debian stable? I think given that rust
> a) is being actively developed b) has it's own tool for installing the latest version - rustup c) lists this tool as the only installation method on their website - we could consider stepping away from supporting the packaged version.

Yes, I want to support Rust from Debian stable. (Rust isn't special. All languages I know are actively developed, be it C++ with gcc or clang, Python with CPython or Rust. I want to be able to use Rust without using some dedicated Rust installer, I prefer using package sources.)

heinrich5991 commented on tools/src/bin/packet_decode.rs:
```diff
@@ -16,11 +16,11 @@ use net::protocol::ChunksIter;
 use net::protocol::Packet;
 use packer::Unpacker;
-use std::fmt;
+
```
Minor: The two new empty lines here could be removed.

heinrich5991 reviewed (commented):
Thanks for working towards a warning-free build. (I guess we could even let the CI deny warnings to make sure no new ones show up.)

However, I'm not willing to increase the minimum supported Rust version just yet, see the other comment. What do you think?

heinrich5991 commented on stats_browser/src/stats_browser.rs:
```diff
@@ -339,7 +339,7 @@ impl<'a> StatsBrowser<'a> {
     }
     fn pump_network(&mut self) {
-        let mut buffer: [u8; 2048] = unsafe { mem::uninitialized() };
+        let mut buffer: [u8; 2048] = unsafe { mem::MaybeUninit::uninit().assume_init() };
```
In every place but here, I think I'd be fine with replacing `mem::uninitialized()` with `mem::zeroed()`.

For now, I can't find a way to keep these bytes uninitialized to pass them to the socket reading; the standard library does exactly what we do here while also stating that it is undefined behavior. I think I'd prefer keeping it as it is until we find a way to pass uninitialized bytes to the OS socket.

Fireball-Teeworlds replied:
It seems that `[0u8; 2048]` should be ok: https://godbolt.org/z/v99MoY.

By the way, it looks like both `mem::zeroed()` and `mem::uninitialized()` result in two calls to memcpy, while `[0u8; 2048]` is just one memset. Maybe just go for the latter one everywhere? :)

Fireball-Teeworlds replied:
Oh, with `-C opt-level=3` `mem::zeroed()` and `[0u8; 2048]` become identical and `mem::uninitialized()` becomes a no-op: https://godbolt.org/z/Yc66d7.

heinrich5991 replied:
https://godbolt.org/z/bEE458

You can see here that the `memset` is still present even if we read from a socket. I don't want to pay a performance cost for not using C at that place.

Fireball-Teeworlds replied:
It seems that the cost isn't high though: memset() for 2048 bytes takes just 39 nanoseconds on my machine: https://gist.github.com/Fireball-Teeworlds/21f126b8dbb285baf88674e91d1fca5f.

I've measured that we call pump_network() around 2000 times within 5-second period (after just starting stats_browser). That should amount to just 0.000078s or 0.0016% of CPU within those 5 seconds.

I've also created a flamegraph of a release build and couldn't find any trace of memset at all (github doesn't allow to attach the original svg :/).

![flamegraph](https://user-images.githubusercontent.com/68162181/91905878-1c7b0180-ec9f-11ea-99d3-b99955bd2e98.png)

Fireball-Teeworlds replied:
I've updated commits according to this finding - in case you'd also find the trade-off okay.

heinrich5991 replied:
I could not verify that the `memset` is called in the optimized variant of your benchmark program, neither via gdb nor via reading the assembly.

In the debug version, it outputs values below 250ns. I guess it's fine…

I'm not convinced by the flamegraph, I doubt memset will get called (unless you verified it for that release build as well), so it could be hiding somewhere.

Fireball-Teeworlds replied:
True, I was testing with debug version (I think I'd have to use `cargo run --release` to test with release).

The flamegraph is for the stats_browser - so there are three options:
1. memset call is so tiny that there are no samples for it (which is likely given that the probability is just 0.0016% for ~200 samples).
2. memset was optimized out (that would be the ideal scenario).
3. memset was inlined. I believe it would've been visible as an empty space above the function if this were the case and it was taking noticeable time.

Fireball-Teeworlds commented:
Changelog:
1. Replaced `mem::uninitialized()` with `mem::zeroed()` and `[0u8; size]`-like initializers.
2. Removed commit that was removing 1.34.2 from build matrix.
3. Fixed that empty line that was added by `cargo fix`.
4. Added auto-generated commit to Cargo.lock (it gets added automatically now unless it's already present).
